File: linnea/code_generation/experiments/templates/Matlab/runner.m
import MatrixGenerator.*;
maxNumCompThreads({num_threads})
% hardcoded directory with naive and recommended implementation
algorithms_dir = fullfile(fileparts(mfilename('fullpath')), 'reference');
addpath(algorithms_dir);
matrices = operand_generator();
naive_ = @() naive(matrices{{:}});
recommended_ = @() recommended(matrices{{:}});
naive_mat = naive_();
recomd_mat = recommended_();
%fprintf('Norm(naive - recomd) %f\\n', norm(naive_mat - recomd_mat));
%fprintf('Naive(0, 0): %f\\n', naive_mat(1, 1));
%fprintf('Naive(0, 0): %f\\n', recomd_mat(1, 1));

benchmarker = Benchmarker();
benchmarker.benchmark('naive_matlab {num_threads}', 20, naive_);
benchmarker.benchmark('recommended_matlab {num_threads}', 20, recommended_);
benchmarker.save('matlab_results_{experiment_name}.txt');

File: linnea/code_generation/experiments/templates/Julia/runner.jl
using Test
using Logging
using MatrixGenerator

using LinearAlgebra.BLAS
BLAS.set_num_threads({num_threads})

include("operand_generator.jl")
{includes}
include("reference/naive.jl")
include("reference/recommended.jl")

function main()
    matrices = operand_generator()

    @info("Performing Test run...")
    result_naive = collect(naive(map(copy, matrices)...)[1])
    result_recommended = collect(recommended(map(copy, matrices)...)[1])
    test_result = isapprox(result_naive, result_recommended, rtol=1e-3)
    @test test_result
{tests}
    @info("Test run performed successfully")

    n = 2000
    rand(n, n)*rand(n, n) # this seems to help to reduce some startup noise

    @info("Running Benchmarks...")
    plotter = Benchmarker.Plot("julia_results_{experiment_name}", ["algorithm"; "threads"]);
{measurements}
    Benchmarker.add_data(plotter, ["naive_julia"; {num_threads}], Benchmarker.measure({repetitions}, naive, matrices...) );
    Benchmarker.add_data(plotter, ["recommended_julia"; {num_threads}], Benchmarker.measure({repetitions}, recommended, matrices...) );
    Benchmarker.finish(plotter);
    @info("Benchmarks complete")
end

main()
